aws/hll/aws-s3-transfer-manager/src/io/part_reader.rs

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+use std::cmp;
+use std::ops::DerefMut;
+use std::sync::Mutex;
+
+use bytes::{Buf, Bytes, BytesMut};
+
+use crate::io::error::Error;
+use crate::io::path_body::PathBody;
+use crate::io::stream::RawInputStream;
+use crate::io::InputStream;
+use crate::MEBIBYTE;
+
+/// Builder for creating a `ReadPart` implementation.
+#[derive(Debug)]
+pub(crate) struct PartReaderBuilder {
+    stream: Option<RawInputStream>,
+    part_size: usize,
+}
+
+impl PartReaderBuilder {
+    pub(crate) fn new() -> Self {
+        Self {
+            stream: None,
+            part_size: 5 * MEBIBYTE as usize,
+        }
+    }
+
+    /// Set the input stream to read from.
+    pub(crate) fn stream(mut self, stream: InputStream) -> Self {
+        self.stream = Some(stream.inner);
+        self
+    }
+
+    /// Set the target part size that should be used when reading data.
+    ///
+    /// All parts except for possibly the last one should be of this size.
+    pub(crate) fn part_size(mut self, part_size: usize) -> Self {
+        self.part_size = part_size;
+        self
+    }
+
+    pub(crate) fn build(self) -> impl ReadPart {
+        let stream = self.stream.expect("input stream set");
+        match stream {
+            RawInputStream::Buf(buf) => {
+                PartReader::Bytes(BytesPartReader::new(buf, self.part_size))
+            }
+            RawInputStream::Fs(path_body) => {
+                PartReader::Fs(PathBodyPartReader::new(path_body, self.part_size))
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+enum PartReader {
+    Bytes(BytesPartReader),
```

Comment:
Does the `Bytes` enum member always refer to bytes in memory or does this also apply to byte streams that can't be "restarted"?

Reply:
right now we only represent two input streams (1) in-memory (Bytes and all it's various conversions to it) and (2) path based. 

I expect we'll expand to at least support `dyn io::Read` or something for any abstract stream as well as some others Micheal mentioned to me around users wanting precise control over how memory is handled or checksums but for now I just implemented these two. I do wonder what a generic approach you and I were discussing might look like still but haven't dug into it.